### PR TITLE
No Map

### PR DESCRIPTION
**Pull request by scottsfarley93**
This PR allows using the mapbox-gl-geocoder plugin without a map, as requested in https://github.com/mapbox/mapbox-gl-geocoder/issues/270. 

Currently, the plugin is added to the map using the map's `addControl` method. In order to facilitate adding the map to other elements, I added a new `addTo` method on the geocoder class. If the addTo method is called with a mapbox-gl-js map, it will be added like normal (the map's addControl method will be called under the hood). If, on the other hand, `geocoder.addTo` is called with a string that looks like a reference to an HTML class or ID, it will add the search bar to that element. In these cases, the geocoder will have no `_map` property and all map-related methods (like updateProximity or flyTo result) will not be executed. 

This still needs tests and I want to make sure there's no edge cases where the lack of a map causes unexpected problems. 

I will also need to:
- return the debug page to it's original state
- Add a new mapbox-gl-js example that shows how to use the geocoder without the map 

\cc @kbauhausness @shabnomnom 

### Launch Checklist

 - [x] briefly describe the changes in this PR
 - [x] write tests for all new functionality
 - [x] run `npm run docs` and commit changes to API.md
 - [x] update CHANGELOG.md with changes under `master` heading before merging

### CODE REVIEW

File: debug/index.js [hunk 1]
```diff
@@ -18,19 +18,29 @@ insertCss(
...
-mapDiv.style.left = 0;
-mapDiv.style.bottom = 0;
-
-var map = new mapboxgl.Map({
-  container: mapDiv,
-  style: 'mapbox://styles/mapbox/streets-v9',
-  center: [-79.4512, 43.6568],
-  zoom: 13
-});
+// var mapDiv = document.body.appendChild(document.createElement('div'));
```
Comment (andrewharvey) [inline debug/index.js]: What about creating a different debug page such as `debug/mapless.js` so both can be tested side by side?

Comment (scottsfarley93) [reply debug/index.js]: Yeah great idea, I'll add that to this PR.

**Comment by scottsfarley93** [comment]
Tapping @yuletide for review.

**Comment by andrewharvey** [comment]
Is it worth linking to the Mapbox TOS obligations when using places? [https://www.mapbox.com/legal/tos#[GAGA]](https://www.mapbox.com/legal/tos#[GAGA]), since you wouldn't be able to use no map unless you exclude places.

File: lib/events.js [hunk 2]
```diff
@@ -85,13 +85,16 @@ MapboxEventManager.prototype = {
...
+
     //pass invalid event
     if (!keyEvent.key) return;
     // don't send events for keys that don't change the input
     // TAB, ESC, LEFT, RIGHT, ENTER, UP, DOWN
     if (keyEvent.metaKey || [9, 27, 37, 39, 13, 38, 40].indexOf(keyEvent.keyCode) !== -1) return;
     var payload = this.getEventPayload('search.keystroke', geocoder);
     payload.lastAction = keyEvent.key;
+    console.log("action: ", payload.lastAction)
+    console.log(payload)
```
Comment (yuletide) [inline lib/events.js]:
```suggestion
```

File: lib/index.js [hunk 3]
```diff
@@ -90,8 +90,43 @@ MapboxGeocoder.prototype = {
...
+      container.addControl(this)
+    }
+    // if the container is not a map, but an html element, then add the control to that element
+    else if (typeof container == 'string' && (container.startsWith('#') || container.startsWith('.'))){
+      const parent = document.querySelectorAll(container);
+      if (parent.length == 0){
+        throw new Error("Element ", container, "not found.")
+      }
+
+      parent.forEach(function(parentEl){
```
Comment (yuletide) [inline lib/index.js]: I was able to create two geocodes with the same class applied to two different elements, but they behave unpredictably once created 
gif

File: lib/index.js [hunk 4]
```diff
@@ -182,20 +217,20 @@ MapboxGeocoder.prototype = {
...
     this.mapMarker = null;
     this._handleMarker = this._handleMarker.bind(this);
-
-    this._mapboxgl = this.options.mapboxgl;
-    if (!this._mapboxgl && this.options.marker) {
-      console.error("No mapboxgl detected in options. Map markers are disabled. Please set options.mapboxgl.");
-      this.options.marker = false;
+    if (this._map){
+      this._mapboxgl = this.options.mapboxgl;
+      if (!this._mapboxgl && this.options.marker) {
```
Comment (yuletide) [inline lib/index.js]:
```suggestion
      if (!this._mapboxgl && this.options.marker) {
       // eslint-disable-next-line no-console
```

File: lib/index.js [hunk 5]
```diff
@@ -182,20 +217,20 @@ MapboxGeocoder.prototype = {
...
     this._typeahead.getItemValue = this.options.getItemValue;
 
-    if (this.options.trackProximity) {
+    if (this.options.trackProximity && this._map) {
       this._updateProximity();
       this._map.on('moveend', this._updateProximity);
     }
 
     this.mapMarker = null;
     this._handleMarker = this._handleMarker.bind(this);
```
Comment (yuletide) [inline lib/index.js]: This would never be called if there's no map right? I'd move this stuff to the if (this._map) block below for simplicity. Also means you don't need to check for map in the handler

File: debug/nomap.js [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+);
+
+var MapboxGeocoder = require('../');
+
+var notMapDiv = document.body.appendChild(document.createElement('div'));
+notMapDiv.style.position = 'absolute';
+notMapDiv.style.top = 0;
+notMapDiv.style.right = 0;
+notMapDiv.style.left = 0;
+notMapDiv.style.bottom = 0;
```
Comment (yuletide) [inline debug/nomap.js]: To reproduce the double stuff add in:
```suggestion
notMapDiv.style.bottom = '50%';

var anotherNotMapDiv = document.body.appendChild(document.createElement('div'));
anotherNotMapDiv.style.position = 'absolute';
anotherNotMapDiv.style.top = '50%';
anotherNotMapDiv.style.right = 0;
anotherNotMapDiv.style.left = 0;
anotherNotMapDiv.style.bottom = 0;
anotherNotMapDiv.style.backgroundColor = 'green';
anotherNotMapDiv.classList.add('notAMap');
```

File: lib/index.js [hunk 7]
```diff
@@ -90,8 +90,43 @@ MapboxGeocoder.prototype = {
...
+   * Add the geocoder to a container. The container can be either a mapboxgl.Map or a reference to an HTML class or ID. 
+   * 
+   * If the container is a mapboxgl.Map, this function will behave identically to Map.addControl(geocoder)
+   * If the container is an html id or class, the geocoder will be appended to that element
+   * 
+   * This function will throw an error if the container is not one a map or a class/id reference
+   * It will also throw an error if the referenced html element cannot be found in the document.body
+   * @param {String|mapboxgl.Map} container A reference to the container to which to add the geocoder 
+   */
+  addTo: function(container){
```
Comment (yuletide) [inline lib/index.js]: Is this the new recommended way to invoke the geocoder? If so we should update the debug/index.js and docs to recommend using this method first, and not `onAdd` (which I think was always awkward so this is a welcome change)

File: lib/index.js [hunk 8]
```diff
@@ -90,8 +90,43 @@ MapboxGeocoder.prototype = {
...
+    // if the container is a map, add the control like normal
+    if (container._controlContainer){
+      //  it's a mapbox-gl map, add like normal 
+      container.addControl(this)
+    }
+    // if the container is not a map, but an html element, then add the control to that element
+    else if (typeof container == 'string' && (container.startsWith('#') || container.startsWith('.'))){
```
Comment (yuletide) [inline lib/index.js]: Suggestion: Allow passing in an HTMLElement

`if (container instanceof HTMLElement){parent = container}` and skip the `querySelector` bit

File: lib/index.js (hunk 7 above)
Comment (scottsfarley93) [reply lib/index.js]: I think the recommended method should still be `map.addControl(geocoder)` -- since I believe this will still be the primary usage. We can add an example that uses this method though.

File: debug/nomap.js (hunk 6 above)
Comment (scottsfarley93) [reply debug/nomap.js]: I added a test to `test.ui.js`, so I'm not sure we need to add this to the debug page.

File: lib/index.js (hunk 7 above)
Comment (yuletide) [reply lib/index.js]: Makes sense. So `map.addControl` if you have a map, otherwise use `addTo(container)`

This is the example I had in mind that we could update to use `addTo` instead of `onAdd` since its more logical: https://docs.mapbox.com/mapbox-gl-js/example/mapbox-gl-geocoder-outside-the-map/

**Comment by pauloanalista** [comment]
Does anyone know if this issue has been resolved?

I want to replace Google Maps with MapBox and use a places autocomplete and I don't have a map.

**Comment by scottsfarley93** [comment]
@pauloanalista this feature was released in v4.5.0. See some documentation [here](https://github.com/mapbox/mapbox-gl-geocoder/blob/master/API.md#addto).
